### PR TITLE
Fix wrong assertion

### PR DESCRIPTION
**Pull request by kkmuffme**
Fix https://github.com/vimeo/psalm/issues/9038
Fix https://github.com/vimeo/psalm/issues/9037

### CODE REVIEW

**Comment by orklah** [comment]
Which error do you mean?

**Comment by orklah** [comment]
I'll need some tests for a change on that level, it's quite sensitive code here

File: src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php [hunk 1]
```diff
@@ -3067,6 +3129,7 @@ private static function getFalseEqualityAssertions(
             }
         }
 
+        /** @psalm-suppress LessSpecificReturnStatement */
```
Comment (orklah) [inline src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:3132]: Could you add the shape inferred by Psalm at the end of those suppressions please?

**Comment by orklah** [comment]
@muglug Could you help with this review please?

I'm wary about changes in Algebra classes

File: src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php (hunk 1 above)
Comment (orklah) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:3132]: yes please, there may be a good reason to suppress that but it's basically impossible to review once the suppression has been made

**Comment by muglug** [comment]
A simplified version of #9038 showing the bug is https://psalm.dev/r/9f56fc9775. It also [reproduces in Hakana](https://hakana.dev/#XQAAgACnAAAAAAAAAAAxmwhIY_U8YQ11C644MsIgebpARAQzKBFrc2cZjQGdvTyHJvxW8MPFlLF5Mu4lhGorhGfnsJKkrsFwVzudePznzZqA8Mt17lYX-XTKoBdrGTwCdy2NPSp-CEh6LrydexO6mG9TF8pgMzM8vkAkCPYl23TORKn__l9gAA%3D%3D), a port of Psalm for another language.

I want to have a crack at fixing it in that tool before reviewing this PR to make sure this PR has the most economical fix. Should take a day or so.

**Comment by muglug** [comment]
I think this is a much simpler fix: https://github.com/vimeo/psalm/compare/master...muglug-dont-check-negation

**Comment by muglug** [comment]
I don't believe my PR is "buggy" just because it doesn't fix your issue — it identified [a bunch of logic issues](https://github.com/vimeo/psalm/commit/234787bcd689a55c8fcd69791f3819ed2c0c32ee) in Psalm's own codebase, which this PR did not.

But feel free to rebase this PR and have another go at fixing your issue!

**Comment by muglug** [comment]
> I'm just saying that the issues here https://github.com/vimeo/psalm/issues/5248 are probably caused by your PR

That issue wasn't caused by my PR — that issue was open until yesterday when he closed it by mistake. It's possible a slightly different version of your PR might fix it!

**Pull request closed without merging**